Add HPA support to ChatQnA

byako opened the pull request:
## Description

This PR introduces HPA support to ChatQnA TGI, Embedding and Reranking services based on custom metrics.

eero-t commented on helm-charts/chatqna/values.yaml:
```diff
@@ -6,6 +6,8 @@
 # Declare variables to be passed into your templates.
 
 replicaCount: 1
+horizontalPodAutoscaler:
```
HPA options in all the values files could include following comment:
```suggestion
# Enabling HPA will:
# - Ignore above replica count
# - Overwrite existing PrometheusAdapter "adapter-config" configMap (with ChatQnA specific custom metric queries):
#   https://github.com/kubernetes-sigs/prometheus-adapter/blob/master/deploy/manifests/config-map.yaml
horizontalPodAutoscaler:
```

eero-t replied:
Why drop this comment?
`# - Ignore above replica count, as it will be controlled by HPA`

byako replied:
Unintentional, rebase / merge-conflict-resolve artifact probably.

byako replied:
Ah, I see what you mean. There is no HPA for ChatQnA service itself, only for tgi, embedding and reranking. TGI is not this service, this is an app service, tgi is in common/tgi.

eero-t commented on helm-charts/chatqna/templates/customMetrics.yaml:
```diff
@@ -0,0 +1,53 @@
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+{{- if .Values.horizontalPodAutoscaler.enabled }}
```
In addition to overwriting `adapter-config`, does this also restart PrometheusAdapter so that the new custom metrics are avaiable for HPA?

byako replied:
No, there is no automatic restarting of arbitrary Pods. Also, we cannot guarantee such a Pod or Deployment existing.

eero-t replied:
Ok, I'll provide blurb for Chart README that instructs user to do that, and tells about the gotchas in enabling HPA.

eero-t reviewed (changes requested):
Dependency to Prometheus is needed for metrics monitoring.

Because HPA overwrites / changes several things, it might be better to have it as a separate PR?

eero-t commented on helm-charts/chatqna/templates/horizontalPorAutoscaler.yaml:
```diff
@@ -0,0 +1,50 @@
+    kind: Deployment
+    name: {{ include "chatqna.fullname" . }}
+  minReplicas: 1
+  maxReplicas: 6
```
It's better if user can specify suitable `maxReplicas` value for each HPA controlled deployment in `values.yaml`.

(Especially as none of these deployments specifies their resource usage, nor anti-/affinity rules, so that k8s could avoid scheduling too many instances on the same nodes.)

byako replied:
Done.

eero-t commented on helm-charts/chatqna/templates/horizontalPorAutoscaler.yaml:
```diff
@@ -0,0 +1,50 @@
+        kind: Service
+        name: tgi-svc
+      target:
+        # tgi_request_latency is already average for all the TGI pods,
+        # so this uses Value instead of averageValue.
+        # On ICL Xeon, max TGI queue wait time target = 4s.
```
Let's add link to docs in case user wants to experiment with this (also to other HPA files):
```suggestion
        # tgi_request_latency is average for all the TGI pods. To avoid replica fluctuations when
        # TGI startup + request processing takes longer than HPA evalution period, this uses
        # "Value" (replicas = metric.value / target.value), instead of "averageValue" type:
        #  https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
```

eero-t commented on helm-charts/chatqna/templates/servicemonitor.yaml:
```diff
@@ -0,0 +1,15 @@
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
```
Add some links for the exposed TGI metrics:
```suggestion
#
# Dashboard for the exposed TGI metrics:
# - https://grafana.com/grafana/dashboards/19831-text-generation-inference-dashboard/
# Metric descriptions:
# - https://github.com/huggingface/text-generation-inference/discussions/1127#discussioncomment-7240527

```

(Did you find better description of them any where else.)

byako replied:
I wasn't looking...

eero-t replied:
Sorry, that was not a question but a typo. I meant:  "That bug comment was best TGI metrics description I found"

lianhao commented on helm-charts/common/embedding-usvc/templates/servicemonitor.yaml:
```diff
@@ -0,0 +1,15 @@
+spec:
+  selector:
+    matchLabels:
+      svc: embedding-svc
```
Can we just use {{- include "embedding-usvc.labels" . | nindent 4 }} here instead of adding new serivce labels?

byako replied:
Yes, should be fine.

lianhao commented on helm-charts/common/embedding-usvc/templates/servicemonitor.yaml:
```diff
@@ -0,0 +1,15 @@
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+apiVersion: monitoring.coreos.com/v1
```
we need to make this whole service monitor by using the same switch as the HPA definition

byako commented on helm-charts/chatqna/templates/service.yaml:
```diff
@@ -6,6 +6,7 @@ kind: Service
 metadata:
   name: {{ include "chatqna.fullname" . }}
   labels:
+    svc: tgi-svc
```
Oh, thanks for pointing out, I'll remove the changes from it and add into `common/tgi`.
I'll rename this into chatqna-svc as part of this PR.

byako replied:
No, sorry for confusion, I'll just move the proposed changes to actual tgi. chatqna service does not need HPA as a part of this PR.

byako commented:
Fixed comments, also made ServiceMonitor take name from the Values like Service has it.

eero-t commented on helm-charts/common/embedding-usvc/templates/deployment.yaml:
```diff
@@ -62,6 +65,9 @@ spec:
           resources:
             {{- toYaml .Values.resources | nindent 12 }}
+          {{- if .Values.horizontalPodAutoscaler.enabled }}
+          terminationGracePeriodSeconds: 60
```
```suggestion
          # extra time to finish processing buffered requests before HPA forcibly terminates pod
          terminationGracePeriodSeconds: 60
```

eero-t commented on helm-charts/common/reranking-usvc/templates/deployment.yaml:
```diff
@@ -62,6 +65,9 @@ spec:
           resources:
             {{- toYaml .Values.resources | nindent 12 }}
+          {{- if .Values.horizontalPodAutoscaler.enabled }}
+          terminationGracePeriodSeconds: 60
```
```suggestion
          # extra time to finish processing buffered requests before HPA forcibly terminates pod
          terminationGracePeriodSeconds: 60
```

eero-t commented on helm-charts/common/tgi/templates/deployment.yaml:
```diff
@@ -68,6 +71,9 @@ spec:
           resources:
             {{- toYaml .Values.resources | nindent 12 }}
+          {{- if .Values.horizontalPodAutoscaler.enabled }}
+          terminationGracePeriodSeconds: 120
```
```suggestion
          # extra time to finish processing buffered requests before HPA forcibly terminates pod
          terminationGracePeriodSeconds: 120
```

eero-t commented on helm-charts/common/embedding-usvc/templates/horizontalPodAutoscaler.yaml:
```diff
@@ -0,0 +1,50 @@
+        kind: Service
+        name: tei-embedding-svc
+      target:
+        # embedding_request_latency is already average for all the TEI pods,
+        # so this uses Value instead of averageValue.
+        # On ICL Xeon, max tei-embedding wait time target = 4s.
```
```suggestion
        # embedding_request_latency is average for all TEI pods. To avoid replica fluctuations when
        # TEI startup + request processing takes longer than HPA evaluation period, this uses
        # "Value" (replicas = metric.value / target.value), instead of "averageValue" type:
        #  https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
```

eero-t commented on helm-charts/common/reranking-usvc/templates/horizontalPodAutoscaler.yaml:
```diff
@@ -0,0 +1,50 @@
+        kind: Service
+        name: tei-reranking-svc
+      target:
+        # reranking_request_latency is already average for all the TEI pods,
+        # so this uses Value instead of averageValue.
+        # On ICL Xeon, max tei-reranking wait time target = 4s.
```
```suggestion
        # reranking_request_latency is average for all TEI pods. To avoid replica fluctuations when
        # TEI startup + request processing takes longer than HPA evaluation period, this uses
        # "Value" (replicas = metric.value / target.value), instead of "averageValue" type:
        #  https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
```

eero-t reviewed (commented):
Some comment update proposals.

byako commented:
Added all suggested comments.

eero-t commented:
Something like this could be added to relevant Chart READMEs, e.g. between their Install and Verify sections:
```
## HorizontalPodAutoscaler (HPA) support

`horizontalPodAutoscaler` option enables HPA scaling for the deployment:
https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/

Autoscaling is based on custom application metrics provided through [Prometheus](https://prometheus.io/).

### Pre-conditions

If cluster does not run [Prometheus operator](https://github.com/prometheus-operator/kube-prometheus)
yet, it SHOULD be be installed before enabling HPA, e.g. by using:
https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack

### Gotchas

Why HPA is opt-in:
* Enabling chart `horizontalPodAutoscaler` option will _overwrite_ cluster's current
  `PrometheusAdapter` configuration with its own custom metrics configuration.
  Take copy of the existing one before install, if that matters:
  `kubectl -n monitoring get cm/adapter-config -o yaml > adapter-config.yaml`
* `PrometheusAdapter` needs to be restarted after install, for it to read the new configuration:
  `ns=monitoring; kubectl -n $ns delete $(kubectl -n $ns get pod --selector app.kubernetes.io/name=prometheus-adapter -o name)`
* By default Prometheus adds [k8s RBAC rules](https://github.com/prometheus-operator/kube-prometheus/blob/main/manifests/prometheus-roleBindingSpecificNamespaces.yaml)
  for accessing metrics from `default`, `kube-system` and `monitoring` namespaces.  If Helm is
  asked to install OPEA services to some other namespace, those rules need to be updated accordingly
* Provided HPA rules are examples for Xeon, for efficient scaling they need to be fine-tune for given setup
  (underlying HW, used models, OPEA version etc) 
```

byako commented:
Added HPA sections to `README.md`s, and moved HPA from `reranking-usvc` chart to `teirerank` which seem to be the actual reranking. `embedding-usvc` seems to be `tei-reranking`, so HPA added to it seems to be the correct place.

eero-t commented:
Bit more work required for the README:
* I had a typo in my blurb: `s/ be fine-tune / be fine-tuned /`
* `horizontalPodAutoscaler` option should be added to table in "Values" section of the READMEs
* READMEs "Verify" sections could include following subsection on verifying that data required by HPA is present

------------------

### Verify HPA metrics

To verify that metrics required by `horizontalPodAutoscaler` option work, check that...

Prometheus found the metric endpoints, i.e. last number on the line is non-zero:
```
prom_url=http://$(kubectl -n monitoring get -o jsonpath="{.spec.clusterIP}:{.spec.ports[0].port}" svc/prometheus-k8s)
curl --no-progress-meter $prom_url/metrics | grep scrape_pool_targets.*-svc
```

Prometheus adapter provides custom metrics for their data:
```
kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1 | jq .resources[].name
```

And those custom metrics have valid values for HPA rules:
```
ns=default;  # OPEA namespace
for m in $(kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1 | jq .resources[].name | tr -d '"' | grep namespaces | sed "s%/%/${ns}/metrics/%"); do
  kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1/$m | jq;
done | grep -e metricName -e value
```

NOTE: HuggingFace TGI and TEI services provide metrics endpoint only after they've processed their first request!

eero-t commented:
If we don't care that user cannot directly copy-paste the command, last check could be just:
```
kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1/namespaces/<NAMESPACE>/metrics/<METRIC> | jq
```

eero-t commented on helm-charts/common/embedding-usvc/templates/_helpers.tpl:
```diff
@@ -40,6 +40,7 @@ helm.sh/chart: {{ include "embedding-usvc.chart" . }}
 app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
 {{- end }}
 app.kubernetes.io/managed-by: {{ .Release.Service }}
+svc: {{ include "embedding-usvc.fullname" . }}
```
```suggestion
```

eero-t commented on helm-charts/common/embedding-usvc/templates/servicemonitor.yaml:
```diff
@@ -0,0 +1,17 @@
+spec:
+  selector:
+    matchLabels:
+      svc: {{ include "embedding-usvc.fullname" . }}
```
```suggestion
      app.kubernetes.io/name: {{ include "embedding-usvc.fullname" . }}
```

eero-t commented on helm-charts/common/teirerank/templates/_helpers.tpl:
```diff
@@ -40,6 +40,7 @@ helm.sh/chart: {{ include "teirerank.chart" . }}
 app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
 {{- end }}
 app.kubernetes.io/managed-by: {{ .Release.Service }}
+svc: {{ include "teirerank.fullname" . }}
```
```suggestion
```

eero-t commented on helm-charts/common/teirerank/templates/servicemonitor.yaml:
```diff
@@ -0,0 +1,17 @@
+spec:
+  selector:
+    matchLabels:
+      svc: {{ include "teirerank.fullname" . }}
```
```suggestion
      app.kubernetes.io/name: {{ include "teirerank.fullname" . }}
```

eero-t commented on helm-charts/common/tgi/servicemonitor.yaml:
```diff
@@ -0,0 +1,22 @@
+spec:
+  selector:
+    matchLabels:
+      svc: {{ include "tgi.fullname" . }}
```
```suggestion
      app.kubernetes.io/name: {{ include "tgi.fullname" . }}
```

eero-t commented on helm-charts/common/tgi/templates/_helpers.tpl:
```diff
@@ -40,6 +40,7 @@ helm.sh/chart: {{ include "tgi.chart" . }}
 app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
 {{- end }}
 app.kubernetes.io/managed-by: {{ .Release.Service }}
+svc: {{ include "tgi.fullname" . }}
```
```suggestion
```

eero-t reviewed (changes requested):
Helm charts helpers unconditionally add a selector label for deployments, so `ServiceMonitor`s can use that instead of new `svc` label.

eero-t commented on helm-charts/common/tgi/README.md:
````diff
@@ -39,6 +68,35 @@ curl http://localhost:2080/generate \
+
+```console
+prom_url=http://$(kubectl -n monitoring get -o jsonpath="{.spec.clusterIP}:{.spec.ports[0].port}" svc/prometheus-k8s)
+curl --no-progress-meter $prom_url/metrics | grep scrape_pool_targets.*-svc
````
```suggestion
curl --no-progress-meter $prom_url/metrics | grep scrape_pool_targets.*tgi
```

eero-t reviewed (changes requested):
OPEA has changed the service names so `*-svc` pattern cannot be used to match the relevant one any more, each service needs its own grep pattern for validation.

eero-t commented on helm-charts/common/tgi/values.yaml:
```diff
@@ -7,6 +7,14 @@
+# - Add example HPA scaling rules with thresholds suitable for Xeon deployments
+# - Require custom metrics ConfigMap available in the main application chart
+horizontalPodAutoscaler:
+  enable: false
```
```suggestion
  enabled: false
```

irisdingbj reviewed (commented):
Besides helm-chart, Would you also update the manifests( https://github.com/opea-project/GenAIInfra/tree/main/microservices-connector/config/manifests)  to accomodate this change?

eero-t reviewed (changes requested):
Nowadays Helm charts `Service` declarations use service's name also for port names.

As port name is hard-coded in `Service`s, I'm suggesting same with `ServiceMonitor`s, but I think _both_ could as well switch to using Helm `include` instead...

eero-t commented on helm-charts/common/embedding-usvc/README.md:
In v0.7, inferencing was done by `tei_embedding` service, now it's done by `tei`, _not_ in `embedding-usvc`.  I.e. this HPA scaling should be moved from `embedding-usvc ` to `tei` (along with relevant name changes).

eero-t commented on helm-charts/common/tei/templates/servicemonitor.yaml:
```diff
@@ -0,0 +1,17 @@
+spec:
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: {{ include "tei.fullname" . }}
```
```suggestion
      app.kubernetes.io/name: {{ include "tei.name" . }}
```

eero-t commented on helm-charts/common/tei/templates/horizontalPodAutoscaler.yaml:
```diff
@@ -0,0 +1,51 @@
+        apiVersion: v1
+        # get metric for named object of given type (in same namespace)
+        kind: Service
+        name: tei-embedding-svc
```
```suggestion
        name: {{ include "tei.fullname" . }}
```

eero-t commented on helm-charts/common/teirerank/templates/horizontalPodAutoscaler.yaml:
```diff
@@ -0,0 +1,51 @@
+        apiVersion: v1
+        # get metric for named object of given type (in same namespace)
+        kind: Service
+        name: tei-reranking-svc
```
```suggestion
        name: {{ include "teirerank.fullname" . }}
```

eero-t commented on helm-charts/common/teirerank/templates/servicemonitor.yaml:
```diff
@@ -0,0 +1,17 @@
+spec:
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: {{ include "teirerank.fullname" . }}
```
```suggestion
      app.kubernetes.io/name: {{ include "teirerank.name" . }}
```

eero-t commented on helm-charts/common/tgi/templates/horizontalPorAutoscaler.yaml:
```diff
@@ -0,0 +1,51 @@
+        apiVersion: v1
+        # get metric for named object of given type (in same namespace)
+        kind: Service
+        name: tgi-svc
```
```suggestion
        name: {{ include "tgi.fullname" . }}
```

eero-t commented on helm-charts/common/tgi/templates/servicemonitor.yaml:
```diff
@@ -0,0 +1,22 @@
+spec:
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: {{ include "tgi.fullname" . }}
```
```suggestion
      app.kubernetes.io/name: {{ include "tgi.name" . }}
```

eero-t reviewed (changes requested):
* Object name specified in HPA rules must match the service name (which is `fullname`)
* Helpers use `name` for Selector labels, not `fullname`

irisdingbj commented:
Besides helm-chart, Would you also update the manifests( https://github.com/opea-project/GenAIInfra/tree/main/microservices-connector/config/manifests)  to accomodate this change? 

> > Besides helm-chart, Would you also update the manifests( https://github.com/opea-project/GenAIInfra/tree/main/microservices-connector/config/manifests) to accomodate this change?
> 
> @irisdingbj As HPA support is disabled by default, running `helm-charts/update_manifests.sh` would just add few extra comments to the manifests, nothing else.
> 
> If you meant generating _additional_ set of manifest files for HPA, I think that's a bad idea. User will then miss Pre-conditions and Gotchas documented in the Helm charts READMEs, and does not have options to configure HPA for underlying cloud setup (e.g. to how many replicas each deployment can be scaled, which depends on how many nodes are available).

Can we genereate manifests with the HPA enabled and then place those manifests under  https://github.com/opea-project/GenAIInfra/tree/main/microservices-connector/config/HPA? In this case if user want to enable HPA via manifest, they will have such choice

byako commented:
Added HPA-enabled manifests.

eero-t reviewed (changes requested):
> Added HPA-enabled manifests.

* Custom metric rules file needs to be added (from chatqna)
* Gaudi & Nvidia files should be dropped as they do not include HPA
* README including HPA Pre-conditions, Gotchas and Validation sections would be nice

PS. IMHO manifests could come in separate PR after this has been merged, there's some script for automating their generation, and some feedback on how well Helm'ed HPA works for others.

(Current HPA values are from earlier testing using OPEA code around v0.7 - v0.8 releases, mostly on 1-2 ICL Xeon nodes.)

byako commented:
tgi_gaudi and tgi_nvidia are flavors of tgi manifest with different resource requested, they do have HPA in them.
I'll add custom metrics.

eero-t commented:
> tgi_gaudi and tgi_nvidia are flavors of tgi manifest with different resource requested, they do have HPA in them.

Using CPU specific metrics thresholds and timings for them does not make any sense.

Their instances can get up much faster (load & warm-up models) and process requests _much_ faster than CPU, therefore their scale up and down timings can be clearly shorter, along with metric thresholds.  It just hasn't been tested yet, what they should be.

With much faster startup & processing, `AverageValue` HPA type is likely to be more relevant for them than `Value` type used with slow CPU inferencing.

eero-t commented on helm-charts/chatqna/templates/customMetrics.yaml:
```diff
@@ -0,0 +1,53 @@
+              resource: namespace
+            service:
+              resource: service
+      - seriesQuery: '{__name__="te_request_inference_duration_sum",service="tei-reranking-svc"}'
+        # Average request latency from TEI histograms, over 1 min
+        metricsQuery: 'rate(te_request_inference_duration_sum{service="tei-reranking-svc",<<.LabelMatchers>>}[1m]) / (0.001+rate(te_request_inference_duration_count{service="tei-reranking-svc",<<.LabelMatchers>>}[1m]))'
```
All 3 `service` labels specified here need to match the actual service name, i.e. Helm's `teirerank.fullname` value, but is that available here?

byako replied:
Not really, no. Need to see how the rule can be added to the custom metrics with patch probably.

byako replied:
There might be a way to import a value from dependency chart, I'm testing.

byako replied:
Done.

eero-t commented on helm-charts/chatqna/templates/customMetrics.yaml:
```diff
@@ -0,0 +1,53 @@
+              resource: namespace
+            service:
+              resource: service
+      - seriesQuery: '{__name__="te_request_inference_duration_sum",service="tei-embedding-svc"}'
+        # Average request latency from TEI histograms, over 1 min
+        metricsQuery: 'rate(te_request_inference_duration_sum{service="tei-embedding-svc",<<.LabelMatchers>>}[1m]) / (0.001+rate(te_request_inference_duration_count{service="tei-embedding-svc",<<.LabelMatchers>>}[1m]))'
```
All 3 `service` labels specified here need to match the actual service name, i.e. Helm's `tei.fullname` value, but is that available here?

byako commented:
Alright, I see. Gaudi and NVidia specific tei and tgi manifests deleted.

eero-t reviewed (changes requested):
Service labels need to match also in Helm chart custom metrics file, and in file pre-generated from it.

lianhao commented on helm-charts/common/teirerank/templates/servicemonitor.yaml:
```diff
@@ -0,0 +1,17 @@
+spec:
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: {{ include "teirerank.name" . }}
```
ditto

eero-t replied:
```suggestion
      {{- include "teirerank.selectorLabels" . | nindent 6 }}
```

lianhao commented on helm-charts/common/tgi/templates/servicemonitor.yaml:
```diff
@@ -0,0 +1,22 @@
+spec:
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: {{ include "tgi.name" . }}
```
ditto

eero-t replied:
```suggestion
      {{- include "tgi.selectorLabels" . | nindent 6 }}
```

byako replied:
Done.

lianhao commented on microservices-connector/config/HPA/customMetrics.yaml:
```diff
@@ -0,0 +1,51 @@
+# Copyright (C) 2024 Intel Corporation
```
Could we add the logic to update manifest files for GMC in update_manifests.sh ?

eero-t replied:
Maybe after #337 has been fixed?

byako replied:
I can do it as a separate PR.

lianhao commented on helm-charts/chatqna/values.yaml:
```diff
@@ -7,6 +7,14 @@
+#   for embedding, reranking, tgi services
+# Upstream default configMap:
+#  - https://github.com/kubernetes-sigs/prometheus-adapter/blob/master/deploy/manifests/config-map.yaml
+horizontalPodAutoscaler:
```
Create the switch here and make it default to false is needed. But make customMetrics.yaml working, we need also to turn on the horizontalPodAutoscaler.enabled switch for tgi/tei/teirerank separately.

I would suggest creating a new value files with all the HPA switches settings-on, and add notes in the chatqna README of how to use it. Just like what we did with gaudi-values.

eero-t replied:
Because custom metrics for all 3 common components are currently added in `chatqna`, READMEs for those common components should mention enabling `horizontalPodAutoscaler` in `chatqna` as their Pre-condition.

However, those scaled components are common ones, i.e. this is not ChatQnA specific...  So, what if custom metrics would have its own Helm chart under common/, and the 3 common components would list it as a dependency?

Though best would be if Helm chart for each of the relevant common components could patch the PrometheusAdapter configMap  themselves (add custom metrics / queries related to their HPA rules to it), instead of that configMap being just overwritten. :-/

byako replied:
I've moved `enabled` flag to global section, so it propagates to children / dependency charts now.

eero-t commented on helm-charts/common/tei/templates/servicemonitor.yaml:
```diff
@@ -0,0 +1,17 @@
+spec:
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: {{ include "tei.name" . }}
```
That helper would make monitor match all the labels in the service object, but that should not be necessary.

Looking at the helpers, there's a separate `.selectorLabels` define for this purpose:
```suggestion
      {{- include "tei.selectorLabels" . | nindent 6 }}
```
which is used also be corresponding service: https://github.com/opea-project/GenAIInfra/blob/main/helm-charts/common/tei/templates/service.yaml

byako replied:
Done.

eero-t commented on helm-charts/chatqna/values.yaml:
```diff
@@ -48,3 +48,11 @@ global:
+  #   for embedding, reranking, tgi services
+  # Upstream default configMap:
+  #  - https://github.com/kubernetes-sigs/prometheus-adapter/blob/master/deploy/manifests/config-map.yaml
+  horizontalPodAutoscaler:
```
This will need "HorizontalPodAutoscaler (HPA) support" section in `chatqna` README too.   It can be same as in common components, with first clause updated e.g. to:
```
`horizontalPodAutoscaler` option enables HPA scaling for the TGI and TEI inferencing deployments:
```

Verification section can be omitted I think, it's enough to have it in TGI & TEI READMEs.

Options table entry description could be e.g:
```
HPA autoscaling for the TGI and TEI service deployments based on metrics they provide. See #pre-conditions and #gotchas before enabling!  (If one doesn't want one of them to be scaled, given service `maxReplicas` can be set to `1`)
```

eero-t commented on helm-charts/common/tei/README.md:
```diff
@@ -21,6 +21,35 @@ MODELDIR=/mnt/opea-models
+If cluster does not run [Prometheus operator](https://github.com/prometheus-operator/kube-prometheus)
+yet, it SHOULD be be installed before enabling HPA, e.g. by using:
+https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack
+
```
```suggestion

`horizontalPodAutoscaler` enabled in top level Helm chart depending on this component  (e.g. `chatqna`), so that relevant custom metric queries are configured for PrometheusAdapter.

```

eero-t commented on helm-charts/common/teirerank/README.md:
```diff
@@ -21,6 +21,35 @@ MODELDIR=/mnt/opea-models
+If cluster does not run [Prometheus operator](https://github.com/prometheus-operator/kube-prometheus)
+yet, it SHOULD be be installed before enabling HPA, e.g. by using:
+https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack
+
```
```suggestion

`horizontalPodAutoscaler` enabled in top level Helm chart depending on this component  (e.g. `chatqna`), so that relevant custom metric queries are configured for PrometheusAdapter.

```

eero-t commented on helm-charts/common/tgi/README.md:
```diff
@@ -24,6 +24,35 @@ MODELDIR=/mnt/opea-models
+If cluster does not run [Prometheus operator](https://github.com/prometheus-operator/kube-prometheus)
+yet, it SHOULD be be installed before enabling HPA, e.g. by using:
+https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack
+
```
```suggestion

`horizontalPodAutoscaler` enabled in top level Helm chart depending on this component  (e.g. `chatqna`), so that relevant custom metric queries are configured for PrometheusAdapter.

```

eero-t reviewed (commented):
Added suggestions on documenting the HPA dependencies between `chatqna` and TGI/TEI components.

eero-t commented on helm-charts/chatqna/templates/customMetrics.yaml:
```diff
@@ -0,0 +1,54 @@
+  config.yaml: |
+      rules:
+      - seriesQuery: '{__name__="tgi_request_inference_duration_sum",service="{{ include "tgi.fullname" .Subcharts.tgi }}"}'
+        metricsQuery: 'rate(tgi_request_inference_duration_sum{service="{{ include "tgi.fullname" .Subcharts.tgi }}",<<.LabelMatchers>>}[1m]) / (0.001+rate(tgi_request_inference_duration_count{service="{{ include "tgi.fullname" .Subcharts.tgi }}",<<.LabelMatchers>>}[1m]))'
+        # Average request latency from TGI histograms, over 1 min
+        # (0.001 divider add is to make sure there's always a valid value)
+        metricsQuery: 'rate(tgi_request_inference_duration_sum{<<.LabelMatchers>>}[1m]) / (0.001+rate(tgi_request_inference_duration_count{<<.LabelMatchers>>}[1m]))'
```
Metrics query is specified twice:
```suggestion
        # Average request latency from TGI histograms, over 1 min
        # (0.001 divider add is to make sure there's always a valid value)
        metricsQuery: 'rate(tgi_request_inference_duration_sum{service="{{ include "tgi.fullname" .Subcharts.tgi }}",<<.LabelMatchers>>}[1m]) / (0.001+rate(tgi_request_inference_duration_count{service="{{ include "tgi.fullname" .Subcharts.tgi }}",<<.LabelMatchers>>}[1m]))'
```

eero-t commented on helm-charts/common/tei/README.md:
````diff
@@ -33,11 +62,41 @@ Open another terminal and run the following command to verify the service if wor
+```console
+ns=default;  # OPEA namespace
+url=/apis/custom.metrics.k8s.io/v1beta1;
+for m in $(kubectl get --raw $url | jq .resources[].name | tr -d '"' | grep namespaces | sed "s%/%/${ns}/metrics/%"); do
+  kubectl get --raw $url/$m | jq;
````
```suggestion
for m in $(kubectl get --raw $url | jq .resources[].name | cut -d/ -f2 | sort -u | tr -d '"'); do
  kubectl get --raw $url/namespaces/$ns/metrics/$m | jq;
```

eero-t commented on helm-charts/common/teirerank/README.md:
````diff
@@ -36,11 +65,41 @@ curl http://localhost:2082/rerank \
+```console
+ns=default;  # OPEA namespace
+url=/apis/custom.metrics.k8s.io/v1beta1;
+for m in $(kubectl get --raw $url | jq .resources[].name | tr -d '"' | grep namespaces | sed "s%/%/${ns}/metrics/%"); do
+  kubectl get --raw $url/$m | jq;
````
```suggestion
for m in $(kubectl get --raw $url | jq .resources[].name | cut -d/ -f2 | sort -u | tr -d '"'); do
  kubectl get --raw $url/namespaces/$ns/metrics/$m | jq;
```

eero-t commented on helm-charts/common/tgi/README.md:
````diff
@@ -39,6 +68,35 @@ curl http://localhost:2080/generate \
+```console
+ns=default;  # OPEA namespace
+url=/apis/custom.metrics.k8s.io/v1beta1;
+for m in $(kubectl get --raw $url | jq .resources[].name | tr -d '"' | grep namespaces | sed "s%/%/${ns}/metrics/%"); do
+  kubectl get --raw $url/$m | jq;
````
```suggestion
for m in $(kubectl get --raw $url | jq .resources[].name | cut -d/ -f2 | sort -u | tr -d '"'); do
  kubectl get --raw $url/namespaces/$ns/metrics/$m | jq;
```

eero-t reviewed (commented):
Came up with IHMO slightly more readable command line example than what I had earlier.  But earlier works fine too.

byako commented:
Added suggested fixes, rebased onto latest main.

eero-t reviewed (approved):
Approved, but please also update `helm-charts/chatqna/README.md` as suggested above.

eero-t commented on helm-charts/chatqna/README.md:
```diff
@@ -34,6 +34,38 @@ helm install chatqna chatqna --set global.HUGGINGFACEHUB_API_TOKEN=${HFTOKEN} --
+If cluster does not run [Prometheus operator](https://github.com/prometheus-operator/kube-prometheus)
+yet, it SHOULD be be installed before enabling HPA, e.g. by using:
+https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack
+
+`horizontalPodAutoscaler` enabled in top level Helm chart depending on this component (e.g. `chatqna`),
+so that relevant custom metric queries are configured for PrometheusAdapter.
```
This part was relevant only for subcomponents:
```suggestion
```

eero-t reviewed (approved):
Failing E2E test seems to be network issue:
```
docker push 100.80.243.74:5000/opea/gmcmanager:84507fc42fbae6a2104ce36457d8ddc5b02c4354
The push refers to repository [100.80.243.74:5000/opea/gmcmanager]
...
337cf9c1bd1f: Retrying in 1 second
received unexpected HTTP status: 500 Internal Server Error
```

yongfengdu commented on helm-charts/chatqna/README.md:
```diff
@@ -34,6 +34,35 @@ helm install chatqna chatqna --set global.HUGGINGFACEHUB_API_TOKEN=${HFTOKEN} --
 
 1. Make sure your `MODELDIR` exists on the node where your workload is schedueled so you can cache the downloaded model for next time use. Otherwise, set `global.modelUseHostPath` to 'null' if you don't want to cache the model.
 
+## HorizontalPodAutoscaler (HPA) support
```
This HPA support section is generic enough. I think maybe we can put it in one place instead of copy/pasting it:
https://github.com/opea-project/GenAIInfra/blob/main/helm-charts/README.md

byako replied:
Done. Thanks, @eero-t for preparing patch.

poussa commented:
@irisdingbj can you approve this and add the v0.9 label. Perhaps merge as well since the automatic merge does not happen since the e2e test fails due to network error in GMC.

byako commented:
I squashed whitespace commit from CI into previous commit and force-pushed to trigger new test round.

irisdingbj commented:
@poussa @eero-t 

> @irisdingbj can you approve this and add the v0.9 label. Perhaps merge as well since the automatic merge does not happen since the e2e test fails due to network error in GMC.

All test passed now and merged into main branch.   v0.9 label already added. Please ask @daisy-ycguo for the process to merge into v0.9 release.
